cudax/include/cuda/experimental/__async/sender/atomic_intrusive_queue.cuh

```diff
@@ -0,0 +1,122 @@
+//===----------------------------------------------------------------------===//
+
+#ifndef __CUDAX_ASYNC_DETAIL_ATOMIC_INTRUSIVE_QUEUE
+#define __CUDAX_ASYNC_DETAIL_ATOMIC_INTRUSIVE_QUEUE
+
+#include <cuda/std/atomic>
+
+#include <cuda/experimental/__async/sender/intrusive_queue.cuh>
+#include <cuda/experimental/__detail/config.cuh>
+
+#include <cuda/experimental/__async/sender/prologue.cuh>
+
+namespace cuda::experimental::__async
+{
+template <auto _NextPtr>
+class _CCCL_TYPE_VISIBILITY_DEFAULT __atomic_intrusive_queue;
+
+template <class _Tp, _Tp* _Tp::* _NextPtr>
+class alignas(64) __atomic_intrusive_queue<_NextPtr>
+{
+public:
+  using __node_pointer        = _Tp*;
+  using __atomic_node_pointer = _CUDA_VSTD::atomic<_Tp*>;
+
+  [[nodiscard]]
+  _CUDAX_API auto empty() const noexcept -> bool
+  {
+    return __head_.load(_CUDA_VSTD::memory_order_relaxed) == nullptr;
+  }
+
+  struct _CCCL_TYPE_VISIBILITY_DEFAULT try_push_result
+  {
+    bool __success;
+    bool __was_empty;
+  };
+
+  _CUDAX_API auto try_push_front(__node_pointer __node) noexcept -> try_push_result
+  {
+    _CCCL_ASSERT(__node != nullptr, "Cannot push a null pointer to the queue");
+    __node_pointer __old_head = __head_.load(_CUDA_VSTD::memory_order_relaxed);
+    __node->*_NextPtr         = __old_head;
+    bool __success            = __head_.compare_exchange_strong(__old_head, __node, _CUDA_VSTD::memory_order_acq_rel);
+    if (__success)
+    {
+      __head_.notify_all();
+    }
+    return {__success, __old_head == nullptr};
+  }
+
+  _CUDAX_API auto push_front(__node_pointer __node) noexcept -> bool
+  {
+    _CCCL_ASSERT(__node != nullptr, "Cannot push a null pointer to the queue");
+    __node_pointer __old_head = __head_.load(_CUDA_VSTD::memory_order_relaxed);
+    do
+    {
+      __node->*_NextPtr = __old_head;
+    } while (!__head_.compare_exchange_weak(__old_head, __node, _CUDA_VSTD::memory_order_acq_rel));
+    __head_.notify_all();
+    return __old_head == nullptr;
+  }
+
+  _CUDAX_API void prepend(__intrusive_queue<_NextPtr> __queue) noexcept
+  {
+    if (__queue.empty())
+    {
+      return;
+    }
+
+    __node_pointer __new_head = __queue.front();
+    __node_pointer __tail     = __queue.back();
+    __node_pointer __old_head = __head_.load(_CUDA_VSTD::memory_order_relaxed);
+    __tail->*_NextPtr         = __old_head;
+    while (!__head_.compare_exchange_weak(__old_head, __new_head, _CUDA_VSTD::memory_order_acq_rel))
+    {
+      __tail->*_NextPtr = __old_head;
+    }
+    __queue.clear();
+    __head_.notify_all();
+  }
+
+  _CUDAX_API auto pop_all() noexcept -> __intrusive_queue<_NextPtr>
+  {
+    return __intrusive_queue<_NextPtr>::make(__reset_head());
+  }
+
+  _CUDAX_API auto pop_all_reversed() noexcept -> __intrusive_queue<_NextPtr>
+  {
+    return __intrusive_queue<_NextPtr>::make_reversed(__reset_head());
+  }
+
+  // Wait until the queue has an item in it:
+  _CUDAX_API void wait() noexcept
```

Comment:
I wonder whether this wait() call should be merged in with a blocking pop_all operation to simplify the API. A wait() call is basically always going to be immediately followed by a pop_all() so why not combine the two?

If you do want to have a separate wait() operation I'd suggest renaming it something like wait_until_nonempty() to be clearer about its semantics.